File: 2015 ICPC North American Qualifier Contest/undetected.py
class UnionFind:
    def __init__(self, n):
        self.p = list(range(n))
        self.rank = [0] * n
        self.set_size = [1] * n
        self.num_sets = n
    
    def find_set(self, i):
        if self.p[i] == i:
            return i
        self.p[i] = self.find_set(self.p[i])
        return self.p[i]
    
    def is_same_set(self, i, j):
        return self.find_set(i) == self.find_set(j)
    
    def union_set(self, i, j):
        if not self.is_same_set(i, j):
            self.num_sets -= 1
            x = self.find_set(i)
            y = self.find_set(j)
            
            if self.rank[x] > self.rank[y]:
                self.p[y] = x
                self.set_size[x] += self.set_size[y]
            else:
                self.p[x] = y
                self.set_size[y] += self.set_size[x]
                if self.rank[x] == self.rank[y]:
                    self.rank[y] += 1

def main():
    n = int(input())
    rads = []
    coords = []
    
    for _ in range(n):
        x, y, r = map(int, input().split())
        rads.append(r)
        coords.append((x, y))
    
    uf = UnionFind(n + 2)
    idx = 0
    
    while idx < n:
        # Check left wall collision
        if coords[idx][0] - rads[idx] < 0:
            uf.union_set(0, idx + 2)
        # Check right wall collision
        if coords[idx][0] + rads[idx] > 200:
            uf.union_set(1, idx + 2)
        
        # Check intersection with previous circles
        for j in range(idx):
            dx = coords[idx][0] - coords[j][0]
            dy = coords[idx][1] - coords[j][1]
            r_sum = rads[idx] + rads[j]
            
            if dx * dx + dy * dy < r_sum * r_sum:
                uf.union_set(idx + 2, j + 2)
        
        # Check if left and right walls are connected
        if uf.is_same_set(0, 1):
            break
            
        idx += 1
    
    print(idx)

if __name__ == "__main__":
    main()

File: 2015 ICPC North American Qualifier Contest/undetected.cs
using System;

class UnionFind {
    private int[] p, rank, setSize;
    public int numSets;
    
    public UnionFind(int n) {
        numSets = n;
        setSize = new int[n];
        rank = new int[n];
        p = new int[n];
        for (int i = 0; i < n; i++) {
            p[i] = i;
            setSize[i] = 1;
        }
    }
    
    public int FindSet(int i) {
        return (p[i] == i) ? i : (p[i] = FindSet(p[i]));
    }
    
    public bool IsSameSet(int i, int j) {
        return FindSet(i) == FindSet(j);
    }
    
    public void UnionSet(int i, int j) {
        if (!IsSameSet(i, j)) {
            numSets--;
            int x = FindSet(i);
            int y = FindSet(j);
            
            if (rank[x] > rank[y]) {
                p[y] = x;
                setSize[x] += setSize[y];
            } else {
                p[x] = y;
                setSize[y] += setSize[x];
                if (rank[x] == rank[y])
                    rank[y]++;
            }
        }
    }
}

class Program {
    static void Main() {
        int n = int.Parse(Console.ReadLine());
        int[] rads = new int[n];
        Tuple<int, int>[] coords = new Tuple<int, int>[n];
        
        for (int i = 0; i < n; i++) {
            string[] parts = Console.ReadLine().Split();
            int x = int.Parse(parts[0]);
            int y = int.Parse(parts[1]);
            int r = int.Parse(parts[2]);
            rads[i] = r;
            coords[i] = Tuple.Create(x, y);
        }
        
        UnionFind uf = new UnionFind(n + 2);
        int idx = 0;
        
        while (idx < n) {
            // Check left wall collision
            if (coords[idx].Item1 - rads[idx] < 0)
                uf.UnionSet(0, idx + 2);
            // Check right wall collision
            if (coords[idx].Item1 + rads[idx] > 200)
                uf.UnionSet(1, idx + 2);
            
            // Check intersection with previous circles
            for (int j = 0; j < idx; j++) {
                int dx = coords[idx].Item1 - coords[j].Item1;
                int dy = coords[idx].Item2 - coords[j].Item2;
                int rSum = rads[idx] + rads[j];
                
                if (dx * dx + dy * dy < rSum * rSum)
                    uf.UnionSet(idx + 2, j + 2);
            }
            
            // Check if left and right walls are connected
            if (uf.IsSameSet(0, 1))
                break;
                
            idx++;
        }
        
        Console.WriteLine(idx);
    }
}